pinot-controller/src/main/java/org/apache/pinot/controller/api/resources/PinotSegmentUploadDownloadRestletResource.java

```diff
@@ -403,6 +413,173 @@ private SuccessResponse uploadSegment(@Nullable String tableName, TableType tabl
+        Pair<String, File> encryptionInfo =
+            encryptSegmentIfNeeded(tempDecryptedFile, tempEncryptedFile, encryptSegment,
+                crypterClassNameInHeader, crypterNameInTableConfig, segmentName, tableNameWithType);
+        File segmentFile = encryptionInfo.getRight();
+
+        // Update download URI if controller is responsible for moving the segment to the deep store
+        URI finalSegmentLocationURI = null;
+        if (copySegmentToFinalLocation) {
+          URI dataDirURI = provider.getDataDirURI();
+          String dataDirPath = dataDirURI.toString();
+          String encodedSegmentName = URIUtils.encode(segmentName);
+          String finalSegmentLocationPath = URIUtils.getPath(dataDirPath, rawTableName, encodedSegmentName);
+          if (dataDirURI.getScheme().equalsIgnoreCase(CommonConstants.Segment.LOCAL_SEGMENT_SCHEME)) {
+            segmentDownloadURIStr = URIUtils.getPath(provider.getVip(), "segments", rawTableName, encodedSegmentName);
+          } else {
+            segmentDownloadURIStr = finalSegmentLocationPath;
+          }
+          finalSegmentLocationURI = URIUtils.getUri(finalSegmentLocationPath);
+        }
+        SegmentUploadMetadata segmentUploadMetadata = new SegmentUploadMetadata(segmentDownloadURIStr,
+            sourceDownloadURIStr, finalSegmentLocationURI, segmentSizeInBytes, segmentMetadata, encryptionInfo);
+        segmentUploadMetadataList.add(segmentUploadMetadata);
+        LOGGER.info("Using segment download URI: {} for segment: {} of table: {} (move segment: {})",
+            segmentDownloadURIStr, segmentFile, tableNameWithType, copySegmentToFinalLocation);
+      } catch (Exception ex) {
+        cleanupTempFiles(tempEncryptedFiles, tempDecryptedFiles, tempSegmentDirs);
+        _controllerMetrics.addMeteredGlobalValue(ControllerMeter.CONTROLLER_SEGMENT_UPLOAD_ERROR,
+            segmentUploadMetadataList.size());
+        throw new ControllerApplicationException(LOGGER, "Exception while processing segments to upload: "
+            + ex.getMessage(), Response.Status.INTERNAL_SERVER_ERROR, ex);
+      }
+    }
+
+    try {
+      ZKOperator zkOperator = new ZKOperator(_pinotHelixResourceManager, _controllerConf, _controllerMetrics);
+      zkOperator.completeSegmentsOperations(tableNameWithType, uploadType, enableParallelPushProtection, allowRefresh,
+          headers, segmentUploadMetadataList);
+      return new SuccessResponse("Successfully uploaded segments: " + segmentNames + " of table: "
+          + tableNameWithType);
+    } catch (Exception ex) {
+      _controllerMetrics.addMeteredGlobalValue(ControllerMeter.CONTROLLER_SEGMENT_UPLOAD_ERROR,
+          segmentUploadMetadataList.size());
+      throw new ControllerApplicationException(LOGGER, "Exception while uploading segments: " + ex.getMessage(),
+          Response.Status.INTERNAL_SERVER_ERROR, ex);
+    } finally {
+      cleanupTempFiles(tempEncryptedFiles, tempDecryptedFiles, tempSegmentDirs);
```

Comment:
why not combine the two try-catch blocks into one, and share this `finally { cleanupTempFiles(); } `

btw, would it possible to process all segments inside one temp dir, then we just need to clean this root temp dir up in the end, and no need to track all temp dirs used by segments separately